imports/ui/templates/components/identity/signup/Signup.jsx

```diff
@@ -0,0 +1,145 @@
+import React, { Component, PropTypes } from 'react';
+import { TAPi18n } from 'meteor/tap:i18n';
+
+import { validateEmail } from '/imports/startup/both/modules/validations.js';
+import { validatePassword, validatePasswordMatch, createUser, validateUsername } from '/imports/startup/both/modules/User.js';
+
+import Warning from '../../../widgets/warning/Warning.jsx';
+
+export default class Signup extends Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      invalidUsername: false,
+      repeatedUsername: false,
+      invalidEmail: false,
+      invalidPassword: false,
+      mismatchPassword: false,
+      alreadyRegistered: false,
+    };
+
+    this.handleSignupRender = this.handleSignupRender.bind(this);
+    this.handleFocus = this.handleFocus.bind(this);
+    this.handleBlur = this.handleBlur.bind(this);
+    this.handleSubmit = this.handleSubmit.bind(this);
+    this.handleSignupError = this.handleSignupError.bind(this);
+  }
+
+  handleSignupRender() {
+    this.props.onClick();
+  }
+
+  handleFocus() {
+    this.setState({ alreadyRegistered: false });
+  }
+
+  handleSignupError() {
+    this.setState({ alreadyRegistered: true });
+  }
+
+  handleBlur(event) {
+    if (event.target.value !== '') {
+      switch (event.target.name) {
+        case 'username-signup':
+          const validUsername = validateUsername(event.target.value);
+          this.setState({ invalidUsername: validUsername.valid });
+          this.setState({ repeatedUsername: validUsername.repeated });
+          break;
+        case 'email-signup':
+          const validEmail = validateEmail(event.target.value);
+          this.setState({ invalidEmail: !validEmail });
+          break;
+        case 'password-signup':
+          const validPassword = validatePassword(event.target.value);
+          this.setState({ invalidPassword: !validPassword });
+          if (document.getElementsByName('mismatchPassword')[0].value !== '') {
+            const validPwMatch = validatePasswordMatch(document.getElementsByName('mismatchPassword')[0].value, event.target.value);
+            this.setState({ mismatchPassword: !validPwMatch });
+          }
+          break;
+        case 'mismatchPassword':
+          const validPwMatch = validatePasswordMatch(document.getElementsByName('password-signup')[0].value, event.target.value);
+          this.setState({ mismatchPassword: !validPwMatch });
+          break;
+      }
+    }
+  }
+
+  handleSubmit(event) {
+    const userData = {
+      username: document.getElementsByName('username-signup')[0].value,
+      email: document.getElementsByName('email-signup')[0].value,
+      password: document.getElementsByName('password-signup')[0].value,
+      mismatchPassword: document.getElementsByName('mismatchPassword')[0].value,
+    };
+
```

Comment:
Because I decided to _not_ have `Signup.jsx` depend on Session variables, I take advantage of the async/await feature to update the component's state correctly. In this case `createUser()` is asynchronous, so the warning popup would not display correctly unless we `await` and set the state within the component.